File: skills/video-production/scripts/cue_sheet.py
#!/usr/bin/env python3
"""Turn a media ledger into a cue sheet and a licence audit.

    python cue_sheet.py --ledger ledger.json --project "Name" --out cue-sheet.md
    python cue_sheet.py --ledger ledger.json --audit          # licence problems only

A cue sheet is a deliverable in its own right: broadcasters and rights societies
ask for one, and a client who is buying a video is buying the right to use the
music in it. This reads the ledger `/media-use` writes when it resolves an asset,
and flags the three things that actually cause trouble:

  · a non-commercial licence in a commercial piece
  · an attribution licence with no attribution recorded
  · an asset with no licence recorded at all

A missing licence is not a formatting problem. It is the reason a video gets
pulled after it ships.
"""
import argparse
import json
import sys
from datetime import date
from pathlib import Path

try:
    sys.stdout.reconfigure(encoding="utf-8")
    sys.stderr.reconfigure(encoding="utf-8")
except Exception:
    pass

NON_COMMERCIAL = ("nc", "noncommercial", "non-commercial", "cc-by-nc", "cc by-nc",
                  "research", "scientific", "personal")
ATTRIBUTION = ("by", "cc-by", "cc by", "attribution")


def norm(v):
    return str(v or "").strip().lower()


def load(path):
    raw = json.loads(Path(path).read_text(encoding="utf-8"))
    if isinstance(raw, dict):
        for key in ("assets", "entries", "ledger", "items"):
            if key in raw and isinstance(raw[key], list):
                return raw[key]
        return list(raw.values()) if all(isinstance(v, dict) for v in raw.values()) else []
    return raw


def audit(rows):
    problems = []
    for r in rows:
        name = r.get("name") or r.get("id") or r.get("path") or "?"
        lic = norm(r.get("licence") or r.get("license"))
        if not lic:
            problems.append(("no licence recorded", name,
                             "Resolve it again through the ledger, or remove the asset."))
            continue
        if any(t in lic for t in NON_COMMERCIAL):
            problems.append(("non-commercial licence", f"{name} — {lic}",
                             "Cannot ship in client or paid work. Replace it."))
        elif any(lic.startswith(t) or t in lic for t in ATTRIBUTION):
            if not (r.get("attribution") or r.get("credit") or r.get("author")):
                problems.append(("attribution required, none recorded", f"{name} — {lic}",
                                 "Record the credit line the licence asks for."))
    return problems


def sheet(rows, project):
    L = [f"# Cue sheet — {project}", "", f"**Prepared:** {date.today().isoformat()}", "",
         "| # | Asset | Type | Use | In | Out | Licence | Source | Credit |",
         "|---:|---|---|---|---|---|---|---|---|"]
    for i, r in enumerate(rows, 1):
        L.append("| {} | {} | {} | {} | {} | {} | {} | {} | {} |".format(
            i,
            r.get("name") or r.get("id") or "",
            r.get("type", ""),
            r.get("use", "background"),
            r.get("in", ""), r.get("out", ""),
            r.get("licence") or r.get("license") or "⚠️ none",
            r.get("source") or r.get("provider") or "",
            r.get("attribution") or r.get("credit") or r.get("author") or ""))
    L += ["", f"{len(rows)} asset(s).", ""]

    problems = audit(rows)
    if problems:
        L += ["## ⚠️ Licence problems", "", "| Problem | Asset | What to do |", "|---|---|---|"]
        for kind, name, fix in problems:
            L.append(f"| {kind} | {name} | {fix} |")
        L.append("")
    else:
        L += ["All assets carry a recorded licence and any required credit.", ""]
    return "\n".join(L), problems


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--ledger", required=True)
    ap.add_argument("--project", default="Untitled")
    ap.add_argument("--out")
    ap.add_argument("--audit", action="store_true")
    a = ap.parse_args()

    rows = load(a.ledger)
    if not rows:
        print("ledger is empty or in an unrecognised shape", file=sys.stderr)
        return 2

    if a.audit:
        problems = audit(rows)
        if not problems:
            print(f"{len(rows)} asset(s) — no licence problems")
            return 0
        for kind, name, fix in problems:
            print(f"FAIL  {kind}: {name}\n      {fix}")
        return 1

    text, problems = sheet(rows, a.project)
    if a.out:
        Path(a.out).write_text(text, encoding="utf-8")
        print(f"wrote {a.out}  ({len(rows)} assets, {len(problems)} problem(s))")
    else:
        print(text)
    return 1 if problems else 0


if __name__ == "__main__":
    sys.exit(main())

File: skills/video-production/scripts/cue_sheet.mjs
#!/usr/bin/env node
/* Turn a media ledger into a cue sheet and a licence audit.
 *
 *   node cue_sheet.mjs --ledger ledger.json --project "Name" --out cue-sheet.md
 *   node cue_sheet.mjs --ledger ledger.json --audit        # licence problems only
 *
 * A cue sheet is a deliverable in its own right: broadcasters and rights societies
 * ask for one, and a client who is buying a video is buying the right to use the
 * music in it. This reads the ledger `/media-use` writes when it resolves an asset,
 * and flags the three things that actually cause trouble:
 *
 *   · a non-commercial licence in a commercial piece
 *   · an attribution licence with no attribution recorded
 *   · an asset with no licence recorded at all
 *
 * A missing licence is not a formatting problem. It is the reason a video gets
 * pulled after it ships.
 */
import { readFileSync, writeFileSync } from "node:fs";

const NON_COMMERCIAL = ["nc", "noncommercial", "non-commercial", "cc-by-nc", "cc by-nc",
                        "research", "scientific", "personal"];
const ATTRIBUTION = ["by", "cc-by", "cc by", "attribution"];

const norm = (v) => String(v ?? "").trim().toLowerCase();
const today = () => new Date().toLocaleDateString("en-CA");

function load(path) {
  const raw = JSON.parse(readFileSync(path, "utf8"));
  if (raw && typeof raw === "object" && !Array.isArray(raw)) {
    for (const key of ["assets", "entries", "ledger", "items"])
      if (key in raw && Array.isArray(raw[key])) return raw[key];
    const vals = Object.values(raw);
    return vals.every((v) => v && typeof v === "object" && !Array.isArray(v)) ? vals : [];
  }
  return raw;
}

function audit(rows) {
  const problems = [];
  for (const r of rows) {
    const name = r.name || r.id || r.path || "?";
    const lic = norm(r.licence || r.license);
    if (!lic) {
      problems.push(["no licence recorded", name,
                     "Resolve it again through the ledger, or remove the asset."]);
      continue;
    }
    if (NON_COMMERCIAL.some((t) => lic.includes(t))) {
      problems.push(["non-commercial licence", `${name} — ${lic}`,
                     "Cannot ship in client or paid work. Replace it."]);
    } else if (ATTRIBUTION.some((t) => lic.startsWith(t) || lic.includes(t))) {
      if (!(r.attribution || r.credit || r.author))
        problems.push(["attribution required, none recorded", `${name} — ${lic}`,
                       "Record the credit line the licence asks for."]);
    }
  }
  return problems;
}

function sheet(rows, project) {
  const L = [`# Cue sheet — ${project}`, "", `**Prepared:** ${today()}`, "",
             "| # | Asset | Type | Use | In | Out | Licence | Source | Credit |",
             "|---:|---|---|---|---|---|---|---|---|"];
  rows.forEach((r, i) => {
    L.push(`| ${i + 1} | ${r.name || r.id || ""} | ${r.type ?? ""} | ${r.use ?? "background"} `
      + `| ${r.in ?? ""} | ${r.out ?? ""} | ${r.licence || r.license || "⚠️ none"} `
      + `| ${r.source || r.provider || ""} | ${r.attribution || r.credit || r.author || ""} |`);
  });
  L.push("", `${rows.length} asset(s).`, "");

  const problems = audit(rows);
  if (problems.length) {
    L.push("## ⚠️ Licence problems", "", "| Problem | Asset | What to do |", "|---|---|---|");
    for (const [kind, name, fix] of problems) L.push(`| ${kind} | ${name} | ${fix} |`);
    L.push("");
  } else {
    L.push("All assets carry a recorded licence and any required credit.", "");
  }
  return [L.join("\n"), problems];
}

const argv = process.argv.slice(2);
const opt = (n) => { const i = argv.indexOf("--" + n); return i === -1 ? undefined : argv[i + 1]; };

if (!opt("ledger")) { console.error("--ledger is required"); process.exit(2); }

const rows = load(opt("ledger"));
if (!rows || !rows.length) {
  console.error("ledger is empty or in an unrecognised shape");
  process.exit(2);
}

if (argv.includes("--audit")) {
  const problems = audit(rows);
  if (!problems.length) {
    console.log(`${rows.length} asset(s) — no licence problems`);
    process.exit(0);
  }
  for (const [kind, name, fix] of problems)
    console.log(`FAIL  ${kind}: ${name}\n      ${fix}`);
  process.exit(1);
}

const [text, problems] = sheet(rows, opt("project") || "Untitled");
if (opt("out")) {
  writeFileSync(opt("out"), text, "utf8");
  console.log(`wrote ${opt("out")}  (${rows.length} assets, ${problems.length} problem(s))`);
} else {
  console.log(text);
}
process.exit(problems.length ? 1 : 0);
